[SE-0335] Enable explicit existential types.

hborla opened the pull request:
This change enables explicit existential types by default, which changes the representation of existential types to use `ExistentialType`, and allows the `any` keyword in the source language.

Note that many of the conditional `getAs<ExistentialType>()` variable bindings are temporary. After explicit existential types are enabled, existential types will always resolve to `ExistentialType` (regardless of whether or not they're spelled with `any`), and the `isExistentialType()` checks can be changed to `is<ExistentialType>()`. I'm holding off on making that change for now to allow explicit existential types to be turned off if needed.

Resolves: rdar://86347236

hborla committed:
[ClangImporter] When importing ObjC pointer types, wrap protocol
compositions in ExistentialType when explicit existentials are
enabled.

hborla committed:
[Sema] Account for ExistentialType in lookupExistentialConformance.

hborla committed:
[Sema] Resolve @_implements protocols using the generic requirement
type resolver context.

hborla committed:
[Deserialization] Fix deserialization of explicit existential types.

hborla committed:
[Type System] When explicit existential types are enabled, wrap Error
in ExistentialType for the type of error values.

hborla committed:
[GenericSignature] Don't allow conformance requirements with explicit
existential types.

hborla committed:
[Codable] When explicit existential types are enabled, synthesize Codable
initializer requirements with explicit existential Encodable and Decodable
parameters.

hborla committed:
[Interop] Handle existential types in ClangTypeConverter and
the ClangImporter.

hborla committed:
[SILOptimizer] Use the constraint type in ExistentialSpecializer when
finding the protocol decl for an existential type.

hborla committed:
[Type Reconstruction] For now, reconstruct existential types without
explicit ExistentialType.

The ASTDemangler needs to produce ExistentialType based on context, not
unconditionally when producing a protocol composition. This is tricky
because existential types are mangled the same way as other protocol
references, e.g. in conformance requirements. For now, drop the explicit
existentials when reconstructing types.

hborla committed:
[Diagnostics] Stage in the new error message for existential types that
must be marked with 'any' as a warning.

slavapestov commented on lib/AST/GenericSignature.cpp:
```diff
@@ -1344,6 +1344,12 @@ void swift::simple_display(raw_ostream &out, GenericSignature sig) {
 
+Requirement::Requirement(RequirementKind kind, Type first, Type second)
+    : RequirementBase(kind, first, second) {
+  assert(!(kind == RequirementKind::Conformance &&
```
IIRC we create conformance requirements with arbitrary types on the right hand side for invalid code like `func foo<T : Int>(_: T)`. I wonder what will happen if you write `func foo<T : any P>(_: T)`, will it hit this assertion? A better place to check generic signature invariants might be in GenericSignature::verify(), which checks minimal/canonical requirements that are output by the GSB.

hborla replied:
Oh of course, you're totally right. I removed the assertion and instead fixed this in `GenericSignatureBuilder::addTypeRequirement` which is where the current check for non-protocol, non-class constraints is. I also added a test for `<T: any P>`. Thanks!

hborla committed:
[Sema] When explicit existential types are enabled, type witnesses that
resolve to ProtocolType or ProtocolCompositionType are always existential
types.

This can happen because type aliases to protocol constraints can also be
used as type witnesses. The type alias itself is still a constraint,
because it can be used in generic signatures in addition to being used
as an existential type, e.g. in parameter position. So, we need to wrap
the type alias in an ExistentialType during type witness resolution.

hborla committed:
[Type System] With explicit existential types, make sure existential
metatypes always wrap the constraint type directly, and the instance
type of an existential metatype is an existential type.

hborla committed:
[Sema] Implement type join for existential types.

hborla committed:
[ConstraintSystem] Fix a few constraint system corner cases with explicit
existential types.

hborla committed:
[TypeResolver] Fix a few corner cases in type resolution for explicit
existential types.

Use the correct type resolver context when resolving SIL metatypes
and function types, and diagnose an error if an explicit existential
type appears in a protocol composition.

hborla committed:
[Type System] Use the constraint type of an existential type in
various places that expect ProtocolType or ProtocoolCompositionType.

hborla committed:
[ASTPrinter] Add an option that controls whether existential types
are printed with the `any` keyword.

For now, printing `any` is off by default in order to turn on explicit
existential types with minimal changes to the test suite. The option
will also allow control over how existential types are printed in
Swift interfaces.

hborla committed:
[Type System] Enable explicit existential types.

hborla commented:
@swift-ci please test

hborla commented:
@swift-ci please test source compatibility

hborla committed:
[AST] Teach ExistentialType::get to only produce ExistentialType when
explicit existential types are enabled.

hborla committed:
[Diagnostics] Remove the warning for 'any Any' and 'any AnyObject'.

hborla commented:
@swift-ci please test

hborla commented:
@swift-ci please test source compatibility

hborla commented on include/swift/AST/Types.h:
```diff
@@ -2731,6 +2734,8 @@ class ExistentialMetatypeType : public AnyMetatypeType {
     return T->getKind() == TypeKind::ExistentialMetatype;
   }
+
+  Type getExistentialInstanceType();
```
I don't like the existence of this method, but it messes up `TypeWalker` if I make `AnyMetatypeType::getInstanceType()` return `ExistentialType` for existential metatypes. Alternatively, I could add `getRawInstanceType()` or something like that for the few places that actually need the stored `InstanceType`. Really, I'd like to just replace `ExistentialMetatypeType` with `ExistentialType(MetatypeType(...))`

slavapestov replied:
I think your longer-term goal of removing ExistentialMetatypeType is good and in the meantime, this workaround is reasonable.

hborla committed:
[Type Reconstruction] Re-construct existential types with ExistentialType
by propagating whether the decoded type is for a requirement through
TypeDecoder.

hborla commented:
@swift-ci please test

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please test source compatibility

hborla commented:
https://github.com/apple/llvm-project/pull/3777

@swift-ci please test

hborla commented:
apple/llvm-project#3777

@swift-ci please test

hborla commented:
apple/llvm-project#3777

@swift-ci please clean test Linux platform

hborla committed:
[Sema] Fix the source compatibility check for member operators with
existential arguments to not return before checking the second
argument type.

hborla committed:
[ConstraintSystem] Return an existential type in getTypeOfMemberReference
when the member is a typealias to a protocol or composition.

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test macOS platform

ahoppen commented:
This PR caused is causing a few crashes in the stress tester: [SR-15742](https://bugs.swift.org/browse/SR-15742).

AnthonyLatsis commented on lib/ClangImporter/ImportType.cpp:
```diff
@@ -1183,6 +1183,10 @@ namespace {
             }
           }
 
+          if (bridgedType->isConstraintType() &&
```
Can `bridgedType` really happen to be an protocol/composition here, or was this a forethought?

hborla replied:
Yes, it can be `Error`

AnthonyLatsis commented on lib/AST/TypeJoinMeet.cpp:
```diff
@@ -271,6 +272,29 @@ CanType TypeJoin::visitExistentialMetatypeType(CanType second) {
+
+  if (joinInstance->is<ExistentialMetatypeType>() ||
+      joinInstance->isAny() ||
+      joinInstance->isAnyObject())
```
Note: `is<ExistentialMetatypeType>()` seems unnecessary here.

hborla replied:
I believe I removed this code in my follow-up PR but I'm not certain

hborla replied:
Ah, I did remove this bit of code in a follow-up commit in this PR. You're right that it's unnecessary.

AnthonyLatsis replied:
Whoops, sorry, I ought to have checked if it's still there on `main`.
